### PR TITLE
libvirtd service: Move mutable configs to /var

### PR DESCRIPTION
**Pull request by fadenb**
Modifies libvirt package to search for configs in /var/lib and changes
libvirtd service to copy the default configs to the new location.

This enables the user to change e.g. the networking configuration with
virsh or virt-manager and keep those settings.

### CODE REVIEW

**Comment by bjornfor** [comment]
Won't this still risk running with stale data? (If/when libvirt ships new config files, they will not be installed to /var/lib/.)

**Comment by fadenb** [comment]
@bjornfor sorry, forgot to address this:
With this PR additional files will be copied but existing files (possibly modified by a user) will not be changed. The risk of running anything stale (that was not explicitly modified by a user) is low.
Furthermore the last changes on the upstream config files were done over 2 years ago (some even 4 or 6 years ago) so this is a very rare thing.

Is this "good enough"? If not is there some common approach how to solve issues like this in NixOS?

**Comment by bjornfor** [comment]
I do not know of a common approach to solve this :-(

Btw, shouldn't the files be copied to /var/lib/libvirtd/? (Seems they are copied directly to /var/lib/ now.)

**Comment by copumpkin** [comment]
@fadenb can you summarize what libvirtd uses the mutability for? I'd like to be able to turn it off where possible.

**Comment by fadenb** [comment]
@copumpkin Libvirt-clients usually connect to the libvirtd and can manage all of libvirtd's settings during runtime. Libvirtd then saves those settings into several .xml files that were in /etc before and now moved (as they are mutable) to /var.

Some examples are IP-ranges for DHCP-Servers or additional network interfaces for VMs.

**Comment by bjornfor** [comment]
@fadenb: And the current NixOS module just kills those user changes on service restart?

**Comment by fadenb** [comment]
@bjornfor yes, the current one force copies the default files from the nix store over to /etc/ (e.g. causing libvirtd to start a dnsmasq instance serving dhcp which is a common thing to disable) on each service start.

**Comment by copumpkin** [comment]
It seems like a similar distinction to `mutableUsers`. I could envision wanting to impose a declarative configuration on `libvirtd`, but I could also envision people wanting to allow mutability in it. The current PR is allowing the mutability, but is there a good way to just turn off the mutable features and specify those things ahead of time in a NixOS module? Will libvirtd deal gracefully with attempted modifications failing?

**Comment by jgillich** [comment]
Mutability is certainly better than the current "behavior". Can we get this merged already? ;)

**Comment by peti** [comment]
How is `/var/lib` preferable over `/etc` for configuration data? That doesn't seem like an improvement to me.

**Comment by globin** [comment]
You write libvirt configs either with virt-manager in the GUI or using virsh possibly even from remote machines, you don't touch them by hand generally so they technically are state. If there was the possibility to configure everything via a nixos module the situation would maybe be different but there are hundreds of possible configuration options..

**Comment by globin** [comment]
I've merged this since this makes libvirt actually work in a reasonable way, but I'd prefer a declarative config but this can be done similar to the config in the wpa_supplicant module so that users can choose between declarative and stateful config.
